### PR TITLE
[dconf] added MGConfItem::listItems()

### PR DESCRIPTION
**Pull request by CODeRUS**

### CODE REVIEW

File: src/mgconfitem.cpp [hunk 1]
```diff
@@ -171,6 +171,94 @@ QStringList MGConfItem::listDirs() const
...
+            if (v) {
+                g_variant_unref(v);
+            }
+        }
+        g_free ((gpointer)item);
+
+        // If we have error set then dconf_is_key() has returned false so we should be safe here
+        if (error) {
+            qWarning() << "MGConfItem::listItems()" << error->message;
+            g_error_free(error);
```
Comment (thp) [inline src/mgconfitem.cpp]: This and the next line can be replaced with: `g_clear_error(&error);`

File: src/mgconfitem.cpp [hunk 2]
```diff
@@ -171,6 +171,94 @@ QStringList MGConfItem::listDirs() const
...
+
+            children.append(d);
+        }
+
+        g_free ((gpointer)dir);
+
+        // If we have error set then dconf_is_key() has returned false so we should be safe here
+        if (error) {
+            qWarning() << "MGConfItem" << error->message;
+            g_error_free(error);
```
Comment (thp) [inline src/mgconfitem.cpp]: Same here - replace these two lines with `g_clear_error(&error);`

File: src/mgconfitem.cpp [hunk 3]
```diff
@@ -171,6 +171,94 @@ QStringList MGConfItem::listDirs() const
...
+    }
+
+    g_strfreev(dirs);
+
+    return children;
+}
+
+QVariantMap MGConfItem::listValues() const
+{
+
```
Comment (thp) [inline src/mgconfitem.cpp]: Style: Remove empty line here

File: src/mgconfitem.h [hunk 4]
```diff
@@ -122,12 +122,31 @@ class MLITESHARED_EXPORT MGConfItem : public QObject
...
+
+    QStringList listKeys() const;
+
+    /*! Return a map of the keys with it's values inside this directory.
+
+        A directory is a key that has children.  The same key might
+        also have a value, but that is confusing and best avoided.
+    */
+
+    QVariantMap listValues() const;
```
Comment (thp) [inline src/mgconfitem.h]: Can't `listValues()` be done (with possibly slightly worse performance if all values are required, but possibly better performance if only some values are required) on the client side by something like:

```
MGConfItem foo(...);
QVariantMap values;
foreach (const QString &key, foo.listKeys()) {
    values[key] = MGConfItem(key).value();
}
```

This way, getting only values of interest is easily possible:

```
MGConfItem foo(...);
QVariantMap values;
foreach (const QString &key, foo.listKeys()) {
    if (key.contains("something_interesting")) {
        values[key] = MGConfItem(key).value();
    }
}
```

If so, we might want to not provide a `listValues()` function in the API until we are sure this is needed (basically if there are enough use cases where performance would benefit compared to the loop-over-keys-and-fetch-value construct).

**Comment by thp** [comment]
Added some comments. Also, please write a few unit tests to demonstrate how the API works and to have some assumptions about behavior encoded (e.g. does `listKeys()` return the full path to the keys or only the names?).

**Comment by thp** [comment]
@CODeRUS Did you have time to look at this again? See my last comment.

**Comment by thp** [comment]
Bump -- any news here?

**Comment by CODeRUS** [comment]
you can close this pr

**Comment by foolab** [comment]
situation as is: missing tests

**Comment by CODeRUS** [comment]
mlite should die, nemomobile-configuration should provide better dconf classes and qml components, and not depend on mlite.

**Comment by monich** [comment]
mlite can't die, it's a public API (https://github.com/sailfish-sdk/sdk-harbour-rpmvalidator/blob/harbour-qa/allowed_libraries.conf#L41).

**Comment by CODeRUS** [comment]
of course it can't just die, but it should be replaced with other better components. MGConfItem should be replaced by nemomobile configuration classes and dependency on mlite for it should be removed, MDesktopEntry can be separated to standalone.

**Comment by monich** [comment]
It can't be replaced and it can't be removed without breaking binary compatibility with the existing harbour apps. It means that we have to keep on improving mlite, if there's something to improve there. There's no reason to duplicate this functionality elsewhere.

**Comment by CODeRUS** [comment]
then mlite should be frozen and not updated, and nemomobile configuration should not depend on mlite and developed separately, and then enabled to harbour apps.

**Comment by monich** [comment]
There's no need for two different APIs in harbour that do the same thing.

**Comment by CODeRUS** [comment]
so why do peoples developing nemomobile-configuration and not improving mlite?

**Comment by monich** [comment]
Beats me. There could be some historical reasons for that. And I'm not particularly attached to mlite, it's just that I don't like things getting multiplied without necessity.

**Comment by CODeRUS** [comment]
mlite is legacy stuff and shoud be frozen and discontinued in some release. imho.

**Comment by monich** [comment]
Nothing should be frozen and discontinued just because it's more than 2 years old.

**Comment by CODeRUS** [comment]
Okay sorry and forget, i just understand nothing :)

**Comment by foolab** [comment]
nemo configuration plugin is a QML only API. mlite MGConfItem is the only C++ harbour allowed API to access dconf and I am not aware of any plans to deprecate it.

A lot of declarative plugins are built on top of C++ libraries and our case is no different.

**Pull request closed without merging**